### ISSUE TITLE
Question about the implementation of equation(4)

### ISSUE THREAD

**Issue by zhengyuan-xie** [Ze-Yang/LGKD#1]
Thanks for your great work! As a beginner in deep learning & PyTorch, I want to know how to implement the LGKD efficiently(It seems that replacing the corresponding gt class' probability with bg class' probability for each pixel costs lots of time). Is there any efficient PyTorch function(eg. gather) used in the official implementation? Looking for your reply!

**Comment by zhengyuan-xie** [comment]
I use the repeat() in the torch and gt mask to reach the goal, is there any more efficient method?

**Comment by Ze-Yang** [comment]
Hi, thank you for your interest in our work. We will release the codes soon after careful testing. For your reference, I provide our implementation of LGKD part as follows. Hope it helps.

```python
class LabelGuidedKnowledgeDistillationLoss(nn.Module):
    def __init__(self, reduction='mean', alpha=1., prev_kd=10, novel_kd=1):
        super().__init__()
        self.reduction = reduction
        self.alpha = alpha
        self.prev_kd = prev_kd
        self.novel_kd = novel_kd
        print("prev kd: {}\t novel kd: {}".format(self.prev_kd, self.novel_kd))

    def forward(self, new_logits, old_logits, targets):
        targets = targets.clone()
        targets[targets < old_logits.shape[1]] = 0
        new_logits = new_logits.permute(0, 2, 3, 1).reshape(-1, new_logits.shape[1])
        old_logits = old_logits.permute(0, 2, 3, 1).reshape(-1, old_logits.shape[1])
        targets = targets.view(-1)

        ignore_mask = targets != 255
        targets = targets[ignore_mask]
        new_logits = new_logits[ignore_mask]
        old_logits = old_logits[ignore_mask]

        new_cl = new_logits.shape[1] - old_logits.shape[1]

        old_logits = old_logits * self.alpha

        novel_mask = targets >= old_logits.shape[1]
        prev_mask = targets < old_logits.shape[1]

        new_logits_novel = new_logits[novel_mask]
        old_logits_novel = old_logits[novel_mask]

        new_logits_prev = new_logits[prev_mask]
        old_logits_prev = old_logits[prev_mask]

        # kd loss for points belonging to the novel classes in the current step
        old_prob_novel = torch.softmax(old_logits_novel, dim=1)
        old_prob_novel = torch.cat([old_prob_novel,
                                    old_prob_novel.new_zeros(old_prob_novel.shape[0], new_cl)], dim=1)
        old_prob_novel[torch.arange(old_prob_novel.shape[0]), targets[novel_mask]] = old_prob_novel[:, 0]
        old_prob_novel[:, 0] = 0
        # old_prob_novel[torch.arange(old_prob_novel.shape[0]), targets[novel_mask]] = 1
        # old_prob_novel[:, :-new_cl] = 0

        den = torch.logsumexp(new_logits_novel, dim=1)
        log_new_prob_novel = new_logits_novel - den.unsqueeze(dim=1)
        loss_novel = (old_prob_novel * log_new_prob_novel).sum(dim=1)

        # kd loss for points belonging to all the classes (including background) in the last step
        old_prob_prev = torch.softmax(old_logits_prev, dim=1)
        old_prob_prev = torch.cat([old_prob_prev,
                                   old_prob_prev.new_zeros(old_prob_prev.shape[0], new_cl)], dim=1)

        den = torch.logsumexp(new_logits_prev, dim=1)
        log_new_prob_prev = new_logits_prev - den.unsqueeze(dim=1)
        loss_prev = (old_prob_prev * log_new_prob_prev).sum(dim=1)

        loss = torch.cat([self.prev_kd * loss_prev, self.novel_kd * loss_novel], dim=0)

        if self.reduction == 'mean':
            outputs = -torch.mean(loss)
        elif self.reduction == 'sum':
            outputs = -torch.sum(loss)
        else:
            outputs = -loss

        return outputs
```

**Comment by zhengyuan-xie** [comment]
Thanks for your fast reply!

**Comment by zhengyuan-xie** [comment]
Hi, I tried the code on PLOP(Pascal VOC 15-1) and the final mIoU is 57.19, which is lower than 61.1(reported in the paper). I just directly added the LGKD loss to the original PLOP codebase, modified the initial step's batch size from 24 to 16 and set prev_kd=5. Do you have suggestions? Here is my result file: https://drive.google.com/file/d/1AWrgbZOP1PvYSM79im93QU3FuiI0yBXb/view?usp=sharing 
Thanks!

**Comment by Ze-Yang** [comment]
Hi, we have released our full codes on 2D and provided our trained models on step 0. If you find it hard to reproduce our results, you can start with our models on step 0. We found that step 0 model training has some randomness, which indeed affects the final results. Thanks.

**Comment by zhengyuan-xie** [comment]
Thanks again for your great work and released code!
